Dockerfile

```
@@ -101,4 +101,4 @@ WORKDIR ${RUNNER_PATH}
 
 # COMMAND
 ENV IN_DOCKER=1
-CMD ["cml-runner"]
+CMD ["cml"]
```

perhaps something like this?

```suggestion
ENTRYPOINT ["/usr/bin/cml"]
CMD ["runner"]
```

---

Absolutely for using `ENTRYPOINT` as the entry point. 💯  Absolutely not for using `runner` as the default command; it doesn't make any sense (to my mind) and that behavior is not even documented.

---

```suggestion
ENTRYPOINT ["/usr/bin/cml"]
```

---

sure but one thing to note is pple who were overriding CMD with e.g. `/bin/bash` will now get an error. They'll have to switch to using `--entrypoint /bin/bash` instead.

---

https://docs.github.com/en/actions/creating-actions/dockerfile-support-for-github-actions#entrypoint

---

This is wrong @0x2b3bfa0 the default command is the runner in docker not just CML

---

> _The default command is the runner in docker not just CML_

Should we leave `runner` as the default command? Given that this behavior is not documented anywhere, I would prefer to run `cml` by default (displaying the help text) and let users choose any command they want.

What advantage do you see in having `runner` as the default command?

---

> Should we leave runner as the default command? Given that this behavior is not documented anywhere, I would prefer to run cml by default (displaying the help text) and let users choose any command they want.

What advantage do you see in having runner as the default command?

Well, we have even videos on how to launch our runner with docker (recommended for self-hosted non cloud)

---

> _Well, we have even videos on how to launch our runner with docker (recommended for self-hosted non cloud)_

That's a very good point, but, if you mean [this video](https://youtu.be/rVq-SCNyxVc?t=514), it also uses the `dvcorg/cml-py3` container, which is stale and deprecated. Merging this change won't stop the video from working, and if we choose to update the video, we can also include the new invocation.

---

Yes, but I still find better to expose the runner.

---

I, as a user, would rather expect to see the main help string than an obscure error because of the lack of tokens. 🤷🏼 

#### Before

```console
$ docker run --rm --tty --interactive iterativeai/cml
```

#### After

```console
$ docker run --rm --tty --interactive iterativeai/cml
```

Users wanting to call the runner could append `runner` to the command.

```console
$ docker run --rm --tty --interactive iterativeai/cml runner
•••
```

---

🤔 I like it 😃

---

My bigger concern is will `jobs.*.container: docker://iterative/cml` work if entrypoint != (ba)sh?

---

It overrides the entrypoint

---

is there docs for that? In the case of [Docker actions, entrypoints are not overriden by GH](https://docs.github.com/en/actions/creating-actions/dockerfile-support-for-github-actions#entrypoint). Are you saying for [`jobs.*.container`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontainer), entrypoints are overridden? To what? I don't see anything in the docs. I'd suspect [`jobs.*.container.options`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontaineroptions): `--entrypoint /bin/bash` will be required :scream:

---

@casperdcl, I have tested it just now, and it's being overriden. 🤷🏼‍♂️  I can't find any documentation, though.

---

However, that's not the case for GitLab, so I guess that we're going to revert the change:
https://docs.gitlab.com/12.10/ee/ci/docker/using_docker_images.html#overriding-the-entrypoint-of-an-image

---

We can either ask GitLab users to override the entry point 🙀  or just use `CMD` as we were doing.

---

Use CMD, thats the main reason we were doing it

---

Done with https://github.com/iterative/cml/pull/703/commits/78764949ac695db0b600c075c3afba6b99b27623

---

> _However there is a missing part. --help is not by default when the parameters are missing_

It is, for the main `cml` command. Subcommands behave as their legacy counterpart.

---

Solved in the best possible way with https://github.com/iterative/cml/pull/703/commits/29bd034c9475a33c0651838a96e3cc3088c34bbd and https://github.com/iterative/cml/pull/703/commits/ecb11db499cba3acf570d29234952f18c49707f4 following @casperdcl's clever idea.

---

oh no what have i done :see_no_evil:

---

I do not fully understand why this. The machine accepts or ```bash``` or ```cml``` only? @0x2b3bfa0

---

Yes, @DavidGOrtega: this smart `ENTRYPOINT` allows GitLab and other tools depending on `CMD` to use our containers for CI/CD without any additional setting.

---

so i.e. ```docker exec myimage nvidia-smi``` wont work?
Whats the benefit of these vs just only having the entrypoint as it was?

---

`docker exec myimage nvidia-smi` won't work unless we allow that command, but `docker exec myimage sh -c nvidia-smi` always will.

---

Thats not very friendly. I totally dislike this change

---

could change to whitelist-only (instead of `(b?a)?sh` look for `(cml)?(runner|pr|publish|send-(comment|github-check)|tensorboard-dev)`)

---

Or, perhaps, try first system commands and then `cml` commands... 🤔 

```json
["/bin/bash", "-c", "which -- \"$(basename -- \"$0\")\" && exec \"$0\" \"$@\" || exec cml \"$0\" \"$@\""]
```

---

I think Im lost. Previously ENTRYPOINT was also overriden in GL right?
Why are we constraining it this way?

---

> _Previously ENTRYPOINT was also overriden in GL right_

It wasn't, and it isn't: https://github.com/iterative/cml/pull/703#discussion_r701136726. GitLab relies on the default entrypoint being able to call `CMD ["sh", "-c", "···"]`

---

> _Why are we constraining it this way?_

So `cml` runs by default without requiring users' explicit choice on which command to run.

---

But the image is used in the GL workflow properly

---

> But the image is used in the GL workflow properly_

GitLab CI/CD runs images with someting like `docker run image sh -c 'shell_finder'` unless users specify a [custom `entrypoint`](https://docs.gitlab.com/ee/ci/docker/using_docker_images.html#available-settings-for-image) pointing to a shell.

---

So it will work if we still maintain the cml entrypoint right?

---

It will work with the smart entrypoint or with the `["/bin/bash", "-c"]` default entrypoint for the [`ubuntu`](https://hub.docker.com/_/ubuntu) images. Setting the entrypoint to `["/usr/bin/cml"]` would break GitLab unless users provide a [custom `entrypoint`](https://docs.gitlab.com/ee/ci/docker/using_docker_images.html#available-settings-for-image) in their workflows.

---

If we use https://github.com/iterative/cml/pull/703#discussion_r701869835, you'll be able to run any system commands and, additionally, default to `cml`

---

But we had a custom entrypoint previously, why was working?

---

We didn't, as far as I know. What we had was a custom `CMD`

https://github.com/iterative/cml/blob/899c6d2e3eaffc66d12492deffa658e8430f1c66/Dockerfile#L104

---

True! Sorry I as willing to say CMD. Why dont we use CMD as we were doing?

---

```bash
docker run iterativeai/cml cml publish # CMD
docker run iterativeai/cml publish # ENTRYPOINT

---

I fiind this better: 
```
docker run iterativeai/cml cml publish # CMD
```

than this:

```
docker exec myimage sh -c nvidia-smi
```

---

Why not have [the best of both worlds](https://github.com/iterative/cml/pull/703#discussion_r701869835)?

```bash
docker run iterativeai/cml publish
docker run iterativeai/cml nvidia-smi
```

---

Applied with https://github.com/iterative/cml/pull/703/commits/a536f0781e8106a8acb5ce5765a6dc045aeaa078

---

I think that https://github.com/iterative/cml/pull/703/commits/a536f0781e8106a8acb5ce5765a6dc045aeaa078 is misleading. cml does not support those commands, how is the user going to guess that docker hack? He is not going to do that with standalone CML.
I do not think is is a good change. What do you think @casperdcl ?